woocommerce/facebook-for-woocommerce
# Handle and log connection errors

Fixes #1805

### Description
We want to log connection errors, this PR allows `connect.woocommerce.com` to send errors when connection to FB fails and logs them.

### Changes
- defines proxy error (`connect.woocommerce.com`) codes and their messages
https://github.com/woocommerce/facebook-for-woocommerce/blob/2830e71d62cb8a25d0cc910140fac902fa47adc3/includes/Handlers/Connection.php#L121-L135

- Handles connections errors and logs them
- Allows connection errors to be log when debug is disabled **but** the site is not connected (**this is because `debug` mode can't be enabled if the site is not connected**)

### Testing instructions
You need an instance of `https://connect.woocommerce.com/` to test this PR end-to-end, if you don't you can test following these instructions:

- Comment these lines in order to pass _nonce_ validation
https://github.com/woocommerce/facebook-for-woocommerce/blob/2830e71d62cb8a25d0cc910140fac902fa47adc3/includes/Handlers/Connection.php#L289-L291
- Now from your local instance go to URL 'https://HOST/?wc-api=wc_facebook_connect&err=1&err_code=token_permission_missing_ads_read' where HOST is your WC instance URL (`fb-wc.local` in my case)
- This request will be redirected to `/wp-admin/admin.php?page=wc-facebook`
- You'll see an error message on your screen:
<img width="1042" alt="Screen Shot 2021-04-14 at 03 29 13" src="https://user-images.githubusercontent.com/532402/114688114-a5e0c080-9cd1-11eb-8700-648114150c18.png">

- Now go to `WooCommerce > Status > Log` and select the last one with name `facebook_for_woocommerce-connect` and confirm error is being logged:
<img width="841" alt="Screen Shot 2021-04-14 at 03 32 02" src="https://user-images.githubusercontent.com/532402/114688523-ffe18600-9cd1-11eb-9c64-a7dfce4ee22c.png">

- Test changing the query param `err_code=token_permission_missing_ads_read` with other errors like: `business_manager_id_missing`, `page_id_missing`, `user_access_token_missing` and validate that the message is correct one
